### PR TITLE
V1 Consensus Module - First Iteration

### PR DESCRIPTION
**Pull request by Olshansk**
Table of contents:

* [PR Context](#pr-context)
* [Review Flow](#review-flow)
* [Changes](#changes)
* [Testing](#testing)

# PR Context

This PR is meant to be the simplest conversion of the [consensus pre-prototype](https://github.com/pokt-network/consensus-pre-prototype) to a consensus module prototype with unnecessary code removed and cleaned up for the official prototype, but still far from perfect.

Please note that there is a lot of follow-up work (e.g. dynamic validator sets) that is not captured in TODOs throughout the code, since this is meant to only act as an overarching baseline.

This PR is also dependant on the smaller [Libraries for Cryptographic Sortition and VRF ](https://github.com/pokt-network/pocket/pull/37/files) PR.

# Review Flow

Since the size of PRs and git history of outgoing code is currently "in transition", a specific flow of review is recommended:

**Flow for review**:
* [ ] 1. @andrewnguyen22 to pull this branch and attempt a cleanup/refactor the code based on experience from v0.
* [ ] 2. @Olshansk and @andrewnguyen22 review & discuss these changes (potentially as part of a protocol hour or ad-hoc).
* [ ] 3. Other reviewers (@derrandz / @oten91 / @emedrano2112 ) are added to the review.
* [ ] 4. Review comments are left and tended to by the author.

## Open question to the reviewers

This review made some modifications to shared libraries which are outside the scope of the consensus module but necessary for end-to-end functionality. Please leave a comment if this should be split out or left as is.

## Focus for review

After (1) and (2) above are complete, please focus on:
* @andrewnguyen22: Business logic in files matching `hotstuff_*` and `pacemaker_*`.
* @emedrano2112: Feedback / ideas on `hotstuff_test.go` and `pacemaker_test.go`
* @derrandz and/or @oten91: Testing instructions/flow and general code structure.

# Changes

### Consensus Module Changes
* Converted most of the types in the consensus pre-prototype types to protobufs from native Go structures
* Added helpers in `block.go` to interface with the Utility module during block proposal / commit
* Hotstuff
    * Implemented Basic Hotstuff including leader handlers, replica handlers, etc…
    * Basic liveness, safety and metadata checks in place
    * “Threshold signatures” are implemented as a list of individual signatures (i.e. not actual partial signatures)
* Pacemaker
    * Implemented a basic version of of the pacemakers
    * Basic timeouts and interrupts implemented to start the next view change
    * Catch-up for round/steps implemented
    * Basic checks & validation in place
    * Added a debug pacemaker module to control & drive it via the CLI
* Leader Election
    * Added a “make-shift” round robin implementation that is deterministic based on height/step/round
* Testing
    * Implement a basic version for “end-to-end” unit testing framework
    * Added end-to-end unit testing for hotstuff and pacemaker

### General Changes
* Implemented a mocked version of the utility module to be used in localnet testing
* Some changes to high level configs (e.g. `config.go`, `validator.go`, `genesis.go`, `test_state.go`, etc…) to make end-to-end testing functionali

### Configuration Changes
* Added basic Pacemaker configurations in `config.json`
* Remove Pacemaker configs from `genesis.json`
* Removed validator private key from `genesis.json`
* Removed `ConsensusParams` from `Genesis`
* Re-ordered the private keys in `config1.json` through `config4.json` so they are ordered based on the lexographic address (for testing purposes)

### Debugging Changes
* Added the CLI debugger with 4 options: `ResetToGenesis`, `PrintNodeState`, `TriggerNextView`, `TogglePacemakerMode`
* Shared Protobufs to sent debug actions/messages

### Makefile Changes
* General `go_vet`,`go_staticcheck`
* Testing: `test_hotstuff`, `test_pacemaker`
* Added the optional `EXTRA_MSG_FAIL` flag

## Future PR TODOs:
* [ ] Extensive edge-case unit testing. Though end-to-end unit tests were added with the first version of the testing framework, extensive edge-case unit testing were omitted in favor of priorities & time
* [ ] Consolidate common types (e.g. `block_cons_temp.proto`) with utility module
* [ ] Delete or refactor `test_state.go`, `genesis.go`, `config.go` and `validator.go`
* [ ] Add a 1-pager design doc for the consensus module

# Testing

### 1. Checkout the source code

```
$ git clone git@github.com:pokt-network/pocket.git pocket_temp
$ git checkout -b issue28/prototype/consensus remotes/origin/issue28/prototype/consensus
```

### 2. End-to-end unit testing

```
# Update/generate protobufs
$ make protogen_clean && make protogen_local

# Update/generate mocks
$ make mockgen

# Run pacemaker unit tests
$ make test_pacemaker

# Run hotstuff unit tests
$ make test_hotstuff
$ EXTRA_MSG_FAIL=true make test_hotstuff

# Run all unit tests
$ make test_all
```

### 3. Localnet testing

Delete any previous docker state
```
$ make docker_wipe
```

First Shell:
```
$ make compose_and_watch
```

Second Shell:
```
$ make client_start
$ make client_connect
> ResetToGenesis
> PrintNodeState # Check committed height is 0
> TriggerNextView
> PrintNodeState # Check committed height is 1
> TriggerNextView
> PrintNodeState # Check committed height is 2
> TogglePacemakerMode # Check that it’s automatic now
> TriggerNextView # Let it rip!
```

### CODE REVIEW

File: cmd/client/main.go [hunk 1]
```diff
@@ -5,28 +5,69 @@ import (
 	"os"
 
 	"github.com/manifoldco/promptui"
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply cmd/client/main.go:7]: Noted. I'm adding a TODO and will also add an AI in the consensus followup issue.

File: consensus/consensus_tests/hotstuff_test.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+		require.Equal(t, uint64(1), nodeState.Height)
+		require.Equal(t, uint8(consensus.NewRound), nodeState.Step)
+		require.Equal(t, uint8(0), nodeState.Round)
+		require.Equal(t, false, nodeState.IsLeader)
+	}
+	for _, message := range newRoundMessages {
+		P2PBroadcast(t, pocketNodes, message)
+	}
+
+	// Leader election is deterministic for now, so we know its NodeId
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply consensus/consensus_tests/hotstuff_test.go:46]: When we implement leader election, it'll fail and future Olshansky will deal with it :)

I have some ideas around seeding

File: consensus/consensus_tests/hotstuff_test.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	numNodes := 4
+	configs := GenerateNodeConfigs(t, numNodes)
+
+	// Create & start test pocket nodes
+	testChannel := make(modules.EventsChannel, 100)
+	pocketNodes := CreateTestConsensusPocketNodes(t, configs, testChannel)
+	for _, pocketNode := range pocketNodes {
+		go pocketNode.Start()
+	}
+	time.Sleep(10 * time.Millisecond) // Avoids minor race conditions if pocket node has not finished starting/initialization
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply consensus/consensus_tests/hotstuff_test.go:25]: Awesome idea! Just added `StartAllTestPocketNodes` in `utils_test.go`.

**Review by andrewnguyen22 — APPROVED** [review]
